### PR TITLE
Fix optional in InputStream (C#)

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR removes unused methods and updates the generated code to use helper methods defined on InputStream.

### CODE REVIEW

File: csharp/src/Ice/InputStream.cs [hunk 1]
```diff
@@ -622,24 +622,6 @@ public byte readByte()
         }
     }
 
-    /// <summary>
-    /// Extracts an optional byte value from the stream.
-    /// </summary>
-    /// <param name="tag">The numeric tag associated with the value.</param>
-    /// <param name="isset">True if the optional value is present, false otherwise.</param>
-    /// <param name="v">The optional value.</param>
-    public void readByte(int tag, out bool isset, out byte v)
```
Comment (bernardnormier) [inline csharp/src/Ice/InputStream.cs:631]: All the "out bool isset" methods were unused, and there is no advantage in providing them, so I removed them.

File: cpp/src/slice2cs/CsUtil.cpp [hunk 2]
```diff
@@ -1714,25 +1714,29 @@ Slice::CsGenerator::writeOptionalSequenceMarshalUnmarshalCode(
                 }
                 else
                 {
-                    out << nl << "if (" << stream << ".readOptional(" << tag << ", " << getOptionalFormat(seq) << "))";
-                    out << sb;
-                    if (builtin->isVariableLength())
+                    if (isArray)
                     {
-                        out << nl << stream << ".skip(4);";
+                        out << nl << param << " = " << stream << ".read" << func << "Seq(" << tag << ");";
```
Comment (bernardnormier) [inline cpp/src/slice2cs/CsUtil.cpp:1719]: We were not checking for this simple path where we provide helper methods (like `readLongSeq(tag)`) on InputStream. As a result, all the `readXxxSeq(tag)` methods were untested.

File: cpp/src/slice2cs/CsUtil.cpp [hunk 3]
```diff
@@ -1714,25 +1714,29 @@ Slice::CsGenerator::writeOptionalSequenceMarshalUnmarshalCode(
...
                     }
-                    else if (builtin->kind() != Builtin::KindByte && builtin->kind() != Builtin::KindBool)
+                    else
                     {
-                        out << nl << stream << ".skipSize();";
+                        out << nl << "if (" << stream << ".readOptional(" << tag << ", " << getOptionalFormat(seq) << "))";
```
Comment (bernardnormier) [inline cpp/src/slice2cs/CsUtil.cpp:1723]: We still need the old path for "cs:generic" sequences.

File: cpp/src/slice2cs/CsUtil.cpp [hunk 4]
```diff
@@ -1714,25 +1714,29 @@ Slice::CsGenerator::writeOptionalSequenceMarshalUnmarshalCode(
...
+                            out << nl << stream << ".skipSize();";
+                        }
+                        writeSequenceMarshalUnmarshalCode(out, seq, scope, param, marshal, true, stream);
+                        out << eb;
+                        out << nl << "else";
+                        out << sb;
+                        out << nl << param << " = null;";
+                        out << eb;
                     }
-                    string tmp = "tmpVal";
```
Comment (bernardnormier) [inline cpp/src/slice2cs/CsUtil.cpp:1727]: Slightly simplified this code by removing tmpVal.

File: csharp/test/Ice/optional/Test.ice [hunk 5]
```diff
@@ -39,7 +39,7 @@ struct VarStruct
 
 sequence<byte> ByteSeq;
 sequence<bool> BoolSeq;
-sequence<short> ShortSeq;
+["cs:generic:List"] sequence<short> ShortSeq;
```
Comment (bernardnormier) [inline csharp/test/Ice/optional/Test.ice:42]: No cs:generic for sequence of built-in tests was a gap in our testings that revealed some bugs.

File: csharp/test/Ice/optional/Test.ice [hunk 6]
```diff
@@ -91,6 +91,7 @@ class MultiOptional
     optional(28) IntMyInterfacePrxDict imipd;
 
     optional(29) BoolSeq bos;
+    optional(30) Object* plainProxy;
```
Comment (bernardnormier) [inline csharp/test/Ice/optional/Test.ice:94]: Improves code coverage.

File: csharp/test/Ice/optional/AllTests.cs [hunk 7]
```diff
@@ -1962,6 +1963,14 @@ internal static void Populate<T>(T[] arr, T value)
                 }
             }
 
+            internal static void Populate<T>(List<T> list, T value)
```
Comment (pepone) [inline csharp/test/Ice/optional/AllTests.cs:1966]: We can also use `new List<...>(Enumerable.Repeat(value, size));` instead of this helper.

**Review by pepone — APPROVED** [review]
Looks good.

File: cpp/src/slice2cs/CsUtil.cpp [hunk 8]
```diff
@@ -1705,7 +1705,7 @@ Slice::CsGenerator::writeOptionalSequenceMarshalUnmarshalCode(
                     }
                     else
                     {
-                        out << nl << "if (" << param << " is not null";
+                        out << nl << "if (" << param << " is not null)";
```
Comment (externl) [inline cpp/src/slice2cs/CsUtil.cpp:1708]: How did this work before?

**Review by externl — APPROVED** [review]

File: cpp/src/slice2cs/CsUtil.cpp (hunk 8 above)
Comment (bernardnormier) [reply cpp/src/slice2cs/CsUtil.cpp:1708]: It didn't. We didn't catch it because there was no test.

**Pull request merged**